### PR TITLE
git: Use blobless clone instead of shallow clones

### PR DESCRIPTION
**Pull request by ReillyBrogan**
Shallow clones are destructive to the history of the repo, and cannot be used in situations where we build from a git source and want to do cherry-picks from that same git source (for instance in binary repos like linux-firmware).

To fix this use blobless clones instead which gets the full history but only downloads blobs on demand. This is slightly less space efficient but works with essentially all git commands. This also lets us simplify the clone and fetch logic somewhat.

### CODE REVIEW

File: builder/source/git.go [hunk 1]
```diff
@@ -125,19 +115,9 @@ func clone(uri, path, ref string) error {
...
 	}
 
-	// We have to add the tag keyword if the ref is a tag, otherwise
-	// git won't actually fetch the tag
-	if len(ref) != 40 {
-		fetchArgs = append(fetchArgs, "tag")
-	}
-
-	fetchArgs = append(fetchArgs, ref)
-
```
Comment (ReillyBrogan) [inline builder/source/git.go:140]: I'm 90% sure that this part is no longer needed due to not using shallow clones so I removed it. If it turns out to be needed a better alternative would be to add `--tags` to the `git fetch` command instead.

File: builder/source/git.go [hunk 2]
```diff
@@ -77,43 +77,33 @@ func (g *GitSource) submodules() error {
...
-	// Check if the reference is a commit
-	if len(ref) == 40 {
-		// Init a new repository at the checkout path
-		initCmd := exec.Command("git", "init", path)
-
-		if err := initCmd.Run(); err != nil {
-			return err
-		}
+// For some reason git blobless clones create pack files with 600 permissions. These break future operations
+// as those files cannot be read by non-root users. Fix those permissions so things work as they should
```
Comment (silkeh) [inline builder/source/git.go]: You're missing a period here:

```suggestion
// as those files cannot be read by non-root users. Fix those permissions so things work as they should.
```

File: builder/source/git.go [hunk 3]
```diff
@@ -77,43 +77,33 @@ func (g *GitSource) submodules() error {
...
+func clone(uri, path, ref string) error {
+	var cmd *exec.Cmd
 
-		if err := fetchCmd.Run(); err != nil {
-			return err
-		}
+	// Create a blobless clone without checking out a ref
+	initCmd := exec.Command("git", "clone", "--filter=blob:none", "--no-checkout", uri, path)
+	initCmd.Stdout = os.Stdout
+	initCmd.Stderr = os.Stderr
```
Comment (silkeh) [inline builder/source/git.go]: Output of called commands should go to stdout:

```suggestion
	initCmd.Stderr = os.Stdout
```

(the `os.Stderr` below is a bug: #79)

File: builder/source/git.go [hunk 4]
```diff
@@ -77,43 +77,33 @@ func (g *GitSource) submodules() error {
...
-		cmd.Dir = path
-	} else {
-		// Not a git commit, so shallow clone the repo at the reference
-		cmd = exec.Command("git", "clone", "--depth", "1", "--branch", ref, uri, path)
+	if err := initCmd.Run(); err != nil {
+		return err
 	}
 
+	// Checkout the ref we want
+	cmd = exec.Command("git", "checkout", ref)
```
Comment (silkeh) [inline builder/source/git.go]: Should we use the modern `git switch -d` here?

File: builder/source/git.go (hunk 1 above)
Comment (silkeh) [reply builder/source/git.go:140]: I agree with your 90% confidence :wink:

**Review by silkeh — CHANGES_REQUESTED** [review]
Few minor comments and a linting failure, but otherwise LGTM!

**Review by silkeh — APPROVED** [review]
LGTM, thanks!

**Pull request merged**